### PR TITLE
Add backend Service and Cache support for SPIFFE Federation resource

### PR DESCRIPTION
**Pull request by strideynet**
Part of https://github.com/gravitational/teleport/issues/44861
As per https://github.com/gravitational/teleport/pull/43348

Adds the local service implementation for SPIFFE Federation and support for this in the Cache used by the Auth Server.

Notably, I've broken out the cache additions for this resource into it's own file rather than appending to the rather unwieldy `collections.go`. I'm hoping we can kick off a bit of a trend here 😅

### CODE REVIEW

**Comment by strideynet** [comment]
Waiting on https://github.com/gravitational/teleport/pull/45058 to merge to let us tidy this up a little more.

File: lib/services/spiffe_federations.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		ctx context.Context, spiffeFederation *machineidv1.SPIFFEFederation,
+	) (*machineidv1.SPIFFEFederation, error)
+	// DeleteSPIFFEFederation deletes a SPIFFE Federation by name.
+	DeleteSPIFFEFederation(ctx context.Context, name string) error
+}
+
+// MarshalSPIFFEFederation marshals the SPIFFEFederation object into a JSON byte
+// array.
+func MarshalSPIFFEFederation(object *machineidv1.SPIFFEFederation, opts ...MarshalOption) ([]byte, error) {
+	object.Version = types.V1
```
Comment (tigrato) [inline lib/services/spiffe_federations.go:56]: should we reset the value here? If we ever change to v2 but using the same message, this field will be problematic

File: lib/auth/accesspoint/accesspoint.go [hunk 2]
```diff
@@ -94,6 +94,7 @@ type Config struct {
 	SAMLIdPSession          services.SAMLIdPSession
 	SecReports              services.SecReports
 	SnowflakeSession        services.SnowflakeSession
+	SPIFFEFederations       cache.SPIFFEFederationReader
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/accesspoint/accesspoint.go:97]: It's the cache's dependency. In this case, had I used `services.SPIFFEFederations`, I'd be requiring methods which are not required by the cache to function (e.g the write side)

 I'd usually have this kind of interfaces unexported, but, there's a bit of a weird case here where the functionality in `accesspoint` really feels like it belongs elsewhere. It's essentially a weird extension of the `cache` package itself. I'd considered redeclaring `SPIFFEFederationsReader` within `accesspoint`, but, I think that there's a better solution down the road (e.g a `accesspoint.UpstreamReaders` struct which holds the interfaces that the cache uses to populate).

Largely, I'm trying to play with the idea that the `services.FoosService` interfaces are unnecessary and problematic. We've see the symptoms of this across our codebase: weirdly named interfaces with "Ext" or "Internal" and methods that just return NotImplemented hung off "wrapper" structs that embed some other type that only partially implements an interface. 

I thought about having `services.SPIFFEFederationsReader`, but it seems to really be the same problem, we should be aiming to create interfaces where they are consumed, not where they are implemented, and to have these interfaces declare the minimum amount of methods. If we think it's not the right time to start solving this problem, then I'm happy to create `services.SPIFFEFederationsReader` and use that, but I feel like that's just prolonging us ever solving this problem.

File: lib/services/spiffe_federations.go (hunk 1 above)
Comment (strideynet) [reply lib/services/spiffe_federations.go:56]: Hm, I don't really mind, I could just enforce that the Version and Kind is set to something within ValidateSPIFFEFederation.

Comment (strideynet) [reply lib/services/spiffe_federations.go]: a610cbb055

**Review by tigrato — APPROVED** [review]

**Review by marcoandredinis — APPROVED** [review]

**Pull request merged**